### PR TITLE
data types and sgd

### PR DESCRIPTION
**Pull request by db7**

### CODE REVIEW

File: features_test.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package cofire
+
+import (
+	"testing"
+
+	"github.com/facebookgo/ensure"
```
Comment (heltonmarx) [inline features_test.go]: here I would like suggest the `github.com/stretchr/testify/require`, or just use STL.

File: features_test.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	}
+	return f
+}
+
+func TestDot(t *testing.T) {
+	user := makeFeatures([]float64{1.0, 2.0, 3.0})
+	product := makeFeatures([]float64{3.0, 2.0, 1.0})
+
+	score := user.dot(product)
+	ensure.DeepEqual(t, score, 10.0)
```
Comment (heltonmarx) [inline features_test.go]: If you decide to use the STL, here you can do like this:
```go
if score != 10.0 {
  t.Errorf("got score %d, want 10.0", score)
}
```

File: features.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
+package cofire
+
+import (
+	"math/rand"
+)
+
+func NewFeatures(rank int) *Features {
```
Comment (heltonmarx) [inline features.go]: Linter:
exported function NewFeatures should have comment or be unexported.

File: sgd.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import "sync"
+
+// SGD is a helper to apply stochastic gradient descent.
+type SGD struct {
+	Gamma  float64
+	Lambda float64
+	bias   float64
+	bsum   float64
+	bcount int
+	m      sync.RWMutex
```
Comment (heltonmarx) [inline sgd.go]: why do you need this mutex, protecting `bias`, `bsum` and `bcount` only ?

File: features_test.go (hunk 2 above)
Comment (db7) [reply features_test.go]: done

File: features.go (hunk 3 above)
Comment (db7) [reply features.go]: done

File: sgd.go (hunk 4 above)
Comment (db7) [reply sgd.go]: I added some comments to clarify that exposed variables are actually constants and don't need to be protected by the mutex.

File: sgd_test.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package cofire
+
+import (
+	"testing"
+
+	"github.com/facebookgo/ensure"
```
Comment (heltonmarx) [inline sgd_test.go]: You can removed here the ensure and use the STL too.

Comment (db7) [reply sgd_test.go]: true

File: cofire.proto [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  Stage    stage  = 1;
+  Rating   rating = 2;
+  Features f      = 3;
+  uint32   iters  = 4;
+}
+
+// Update messages overwrite the U or P features of in the user/product's
+// entry.
+message Update {
+  Features u = 2;
```
Comment (frairon) [inline cofire.proto]: any reason it starts with 2 instead of 1?

File: features.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+func (f *Features) Randomize() *Features {
+	for i := range f.V {
+		f.V[i] = rand.Float64()
+	}
+	return f
+}
+
+// zeros return true if features are all 0.
+func (f *Features) zeros() bool {
+	if f == nil {
```
Comment (frairon) [inline features.go]: only the `zeros`-function checks for "f" being nil, but no other function does. Is that on purpose?

Comment (db7) [reply features.go]: right, fixed

File: cofire.proto (hunk 6 above)
Comment (db7) [reply cofire.proto]: fixed

**Review by frairon — APPROVED** [review]

**Pull request merged**